ICE: `name of non-Rpitit assoc item`  (episode 2)

the saga continues!
<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_middle/src/ty/assoc.rs:43:25: 'name of non-Rpitit assoc item'', 'thread 'rustc' panicked at compiler/rustc_middle/src/ty/assoc.rs:43:25: 'name of non-Rpitit assoc item''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
trait Foo {
    fn test() -> impl Foo(u32) -> u32 {}
}

````

original:
````rust
trait Foo {
    fn test() -> impl Foo(u32) -> u32 {
        || x.count_ones()
    }
}

````

Version information
````
rustc 1.88.0-nightly (78f2104e3 2025-04-16)
binary: rustc
commit-hash: 78f2104e334068d5a892a170d50193c0025c690e
commit-date: 2025-04-16
host: x86_64-unknown-linux-gnu
release: 1.88.0-nightly
LLVM version: 20.1.2

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/78f2104e334068d5a892a170d50193c0025c690e/compiler/rustc_middle/src/ty/assoc.rs#L37-L49

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.JNaW56ApoC9W/rustc_testrunner_tmpdir_reporting.WdMzLCnsDMQF/mvce.rs:3:2
  |
3 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.JNaW56ApoC9W/rustc_testrunner_tmpdir_reporting.WdMzLCnsDMQF/mvce.rs`

error[E0658]: parenthetical notation is only stable when used with `Fn`-family traits
 --> /tmp/icemaker_global_tempdir.JNaW56ApoC9W/rustc_testrunner_tmpdir_reporting.WdMzLCnsDMQF/mvce.rs:2:23
  |
2 |     fn test() -> impl Foo(u32) -> u32 {}
  |                       ^^^^^^^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(unboxed_closures)]` to the crate attributes to enable
  = note: this compiler was built on 2025-04-16; consider upgrading it if it is out of date


thread 'rustc' panicked at compiler/rustc_middle/src/ty/assoc.rs:43:25:
name of non-Rpitit assoc item
stack backtrace:
   0:     0x7d0bb86e87b3 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hbe9a35f36773aa35
   1:     0x7d0bb8e05b07 - core::fmt::write::hc5bf8d0f3fe78c0c
   2:     0x7d0bba2a6011 - std::io::Write::write_fmt::ha84e19e9665c5650
   3:     0x7d0bb86e8612 - std::sys::backtrace::BacktraceLock::print::h8bb069e50731329e
   4:     0x7d0bb86ebf8a - std::panicking::default_hook::{{closure}}::h9d41ab1a4babb09c
   5:     0x7d0bb86ebb0f - std::panicking::default_hook::h6eac92de6962e409
   6:     0x7d0bb775a5b3 - std[2ad6d0fee2e09edc]::panicking::update_hook::<alloc[9d3eded591b7d627]::boxed::Box<rustc_driver_impl[5fbee50f607e217]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7d0bb86ec803 - std::panicking::rust_panic_with_hook::hfe63a727073f604a
   8:     0x7d0bb86ec4fa - std::panicking::begin_panic_handler::{{closure}}::h92b47807861c6a3c
   9:     0x7d0bb86e8c89 - std::sys::backtrace::__rust_end_short_backtrace::hb9f6a948a925f382
  10:     0x7d0bb86ec1bd - __rustc[9e5fbd37ce37ba6c]::rust_begin_unwind
  11:     0x7d0bb508d1c0 - core::panicking::panic_fmt::hcb046f91e9c65441
  12:     0x7d0bb6393fab - core::option::expect_failed::hb997976540264db6
  13:     0x7d0bb78de395 - <core[58638a26b282635c]::iter::adapters::map::Map<core[58638a26b282635c]::iter::adapters::filter::Filter<core[58638a26b282635c]::iter::adapters::filter::Filter<core[58638a26b282635c]::iter::adapters::filter::Filter<core[58638a26b282635c]::iter::adapters::map::Map<core[58638a26b282635c]::iter::adapters::map::Map<core[58638a26b282635c]::slice::iter::Iter<(core[58638a26b282635c]::option::Option<rustc_span[85c7266a28769ec9]::symbol::Symbol>, rustc_middle[9902b4682aa0495d]::ty::assoc::AssocItem)>, <rustc_data_structures[2fd0664b6811e02d]::sorted_map::index_map::SortedIndexMultiMap<u32, core[58638a26b282635c]::option::Option<rustc_span[85c7266a28769ec9]::symbol::Symbol>, rustc_middle[9902b4682aa0495d]::ty::assoc::AssocItem>>::iter::{closure#0}>, <rustc_middle[9902b4682aa0495d]::ty::assoc::AssocItems>::in_definition_order::{closure#0}>, <rustc_hir_analysis[97d869527a21e1a5]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::get_unbound_associated_types::{closure#0}>, <rustc_hir_analysis[97d869527a21e1a5]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::get_unbound_associated_types::{closure#1}>, <rustc_hir_analysis[97d869527a21e1a5]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::get_unbound_associated_types::{closure#2}>, <rustc_hir_analysis[97d869527a21e1a5]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::get_unbound_associated_types::{closure#3}> as core[58638a26b282635c]::iter::traits::iterator::Iterator>::next
  14:     0x7d0bb793efd7 - <rustc_hir_analysis[97d869527a21e1a5]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs>::get_unbound_associated_types
  15:     0x7d0bb7942be7 - <rustc_hir_analysis[97d869527a21e1a5]::errors::wrong_number_of_generic_args::WrongNumberOfGenericArgs as rustc_errors[fd960650b33fb9c4]::diagnostic::Diagnostic>::into_diag
  16:     0x7d0bb90e431b - <dyn rustc_hir_analysis[97d869527a21e1a5]::hir_ty_lowering::HirTyLowerer>::lower_generic_args_of_path::{closure#0}
  17:     0x7d0bb90dd639 - <dyn rustc_hir_analysis[97d869527a21e1a5]::hir_ty_lowering::HirTyLowerer>::lower_poly_trait_ref
  18:     0x7d0bb90dcba4 - <dyn rustc_hir_analysis[97d869527a21e1a5]::hir_ty_lowering::HirTyLowerer>::lower_bounds::<&[rustc_hir[603e0a66c7ecd982]::hir::GenericBound]>
  19:     0x7d0bb9a40d61 - rustc_hir_analysis[97d869527a21e1a5]::collect::item_bounds::associated_type_bounds
  20:     0x7d0bb9a403d4 - rustc_hir_analysis[97d869527a21e1a5]::collect::item_bounds::explicit_item_bounds
  21:     0x7d0bb9a403a7 - rustc_query_impl[2236df548ee42a79]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2236df548ee42a79]::query_impl::explicit_item_bounds::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 16usize]>>
  22:     0x7d0bb9a4036b - <rustc_query_impl[2236df548ee42a79]::query_impl::explicit_item_bounds::dynamic_query::{closure#2} as core[58638a26b282635c]::ops::function::FnOnce<(rustc_middle[9902b4682aa0495d]::ty::context::TyCtxt, rustc_span[85c7266a28769ec9]::def_id::DefId)>>::call_once
  23:     0x7d0bb9001f10 - rustc_query_system[be2b43304ba7dfb]::query::plumbing::try_execute_query::<rustc_query_impl[2236df548ee42a79]::DynamicConfig<rustc_query_system[be2b43304ba7dfb]::query::caches::DefIdCache<rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[2236df548ee42a79]::plumbing::QueryCtxt, false>
  24:     0x7d0bb98ddea0 - rustc_query_impl[2236df548ee42a79]::query_impl::explicit_item_bounds::get_query_non_incr::__rust_end_short_backtrace
  25:     0x7d0bb9573008 - rustc_hir_analysis[97d869527a21e1a5]::check::compare_impl_item::check_type_bounds
  26:     0x7d0bb9650c98 - rustc_hir_analysis[97d869527a21e1a5]::check::check::check_item_type
  27:     0x7d0bb95f931a - rustc_hir_analysis[97d869527a21e1a5]::check::wfcheck::check_well_formed
  28:     0x7d0bb95f78eb - rustc_query_impl[2236df548ee42a79]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2236df548ee42a79]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 1usize]>>
  29:     0x7d0bb95f7194 - rustc_query_system[be2b43304ba7dfb]::query::plumbing::try_execute_query::<rustc_query_impl[2236df548ee42a79]::DynamicConfig<rustc_data_structures[2fd0664b6811e02d]::vec_cache::VecCache<rustc_span[85c7266a28769ec9]::def_id::LocalDefId, rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[be2b43304ba7dfb]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[2236df548ee42a79]::plumbing::QueryCtxt, false>
  30:     0x7d0bb95f6c83 - rustc_query_impl[2236df548ee42a79]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  31:     0x7d0bb95f30e5 - rustc_hir_analysis[97d869527a21e1a5]::check::wfcheck::check_mod_type_wf
  32:     0x7d0bb95f2edd - rustc_query_impl[2236df548ee42a79]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2236df548ee42a79]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 1usize]>>
  33:     0x7d0bb9991999 - rustc_query_system[be2b43304ba7dfb]::query::plumbing::try_execute_query::<rustc_query_impl[2236df548ee42a79]::DynamicConfig<rustc_query_system[be2b43304ba7dfb]::query::caches::DefaultCache<rustc_span[85c7266a28769ec9]::def_id::LocalModDefId, rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[2236df548ee42a79]::plumbing::QueryCtxt, false>
  34:     0x7d0bb9991710 - rustc_query_impl[2236df548ee42a79]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7d0bb903ffe6 - rustc_hir_analysis[97d869527a21e1a5]::check_crate
  36:     0x7d0bb91d46f2 - rustc_interface[8c35278d6665784]::passes::run_required_analyses
  37:     0x7d0bb99a0dde - rustc_interface[8c35278d6665784]::passes::analysis
  38:     0x7d0bb99a0dad - rustc_query_impl[2236df548ee42a79]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[2236df548ee42a79]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 0usize]>>
  39:     0x7d0bb9e141bd - rustc_query_system[be2b43304ba7dfb]::query::plumbing::try_execute_query::<rustc_query_impl[2236df548ee42a79]::DynamicConfig<rustc_query_system[be2b43304ba7dfb]::query::caches::SingleCache<rustc_middle[9902b4682aa0495d]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[2236df548ee42a79]::plumbing::QueryCtxt, false>
  40:     0x7d0bb9e13eb8 - rustc_query_impl[2236df548ee42a79]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7d0bb9cd223e - rustc_interface[8c35278d6665784]::passes::create_and_enter_global_ctxt::<core[58638a26b282635c]::option::Option<rustc_interface[8c35278d6665784]::queries::Linker>, rustc_driver_impl[5fbee50f607e217]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  42:     0x7d0bb9f0acc4 - rustc_interface[8c35278d6665784]::interface::run_compiler::<(), rustc_driver_impl[5fbee50f607e217]::run_compiler::{closure#0}>::{closure#1}
  43:     0x7d0bb9cf36e8 - std[2ad6d0fee2e09edc]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[8c35278d6665784]::util::run_in_thread_with_globals<rustc_interface[8c35278d6665784]::util::run_in_thread_pool_with_globals<rustc_interface[8c35278d6665784]::interface::run_compiler<(), rustc_driver_impl[5fbee50f607e217]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  44:     0x7d0bb9cf3ff4 - <<std[2ad6d0fee2e09edc]::thread::Builder>::spawn_unchecked_<rustc_interface[8c35278d6665784]::util::run_in_thread_with_globals<rustc_interface[8c35278d6665784]::util::run_in_thread_pool_with_globals<rustc_interface[8c35278d6665784]::interface::run_compiler<(), rustc_driver_impl[5fbee50f607e217]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[58638a26b282635c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  45:     0x7d0bb9cf53f7 - std::sys::pal::unix::thread::Thread::new::thread_start::h7f6705cc982c7917
  46:     0x7d0bb3ca370a - <unknown>
  47:     0x7d0bb3d27aac - <unknown>
  48:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.88.0-nightly (78f2104e3 2025-04-16) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [explicit_item_bounds] finding item bounds for `Foo::{anon_assoc#0}`
#1 [check_well_formed] checking that `Foo` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [explicit_item_bounds] finding item bounds for `Foo::{anon_assoc#0}`
#1 [check_well_formed] checking that `Foo` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_hir_analysis/src/errors/wrong_number_of_generic_args.rs
@@ -486,15 +486,15 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
             let items: &AssocItems = self.tcx.associated_items(self.def_id);
             items
                 .in_definition_order()
-                .filter(|item| item.is_type())
                 .filter(|item| {
-                    !self
-                        .gen_args
-                        .constraints
-                        .iter()
-                        .any(|constraint| constraint.ident.name == item.name())
+                    item.is_type()
+                        && !item.is_impl_trait_in_trait()
+                        && !self
+                            .gen_args
+                            .constraints
+                            .iter()
+                            .any(|constraint| constraint.ident.name == item.name())
                 })
-                .filter(|item| !item.is_impl_trait_in_trait())
                 .map(|item| self.tcx.item_ident(item.def_id).to_string())
                 .collect()
         } else {

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
@@ -29,7 +29,7 @@ use rustc_errors::codes::*;
 use rustc_errors::{
     Applicability, Diag, DiagCtxtHandle, ErrorGuaranteed, FatalError, struct_span_code_err,
 };
-use rustc_hir::def::{CtorKind, CtorOf, DefKind, Namespace, Res};
+use rustc_hir::def::{CtorKind, CtorOf, DefKind, Res};
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::{self as hir, AnonConst, GenericArg, GenericArgs, HirId};
 use rustc_infer::infer::{InferCtxt, TyCtxtInferExt};
@@ -1731,9 +1731,9 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                 tcx.associated_items(*trait_def_id)
                         .in_definition_order()
                         .any(|i| {
-                            i.namespace() == Namespace::TypeNS
+                            i.is_type()
+                                && !i.is_impl_trait_in_trait()
                                 && i.ident(tcx).normalize_to_macros_2_0() == assoc_ident
-                                && i.is_type()
                         })
                     // Consider only accessible traits
                     && tcx.visibility(*trait_def_id)

--- a/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-2.rs
+++ b/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-2.rs
@@ -1,0 +1,12 @@
+trait Foo {
+    fn rpitit() -> impl Sized;
+}
+
+// Ensure that we don't try to probe the name of the RPITIT when looking for
+// fixes to suggest for the redundant generic below.
+
+fn test<T: Foo<i32, Assoc = i32>>() {}
+//~^ ERROR trait takes 0 generic arguments but 1 generic argument was supplied
+//~| ERROR associated type `Assoc` not found for `Foo`
+
+fn main() {}

--- a/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-2.stderr
+++ b/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-2.stderr
@@ -1,0 +1,24 @@
+error[E0107]: trait takes 0 generic arguments but 1 generic argument was supplied
+  --> $DIR/dont-probe-missing-item-name-2.rs:8:12
+   |
+LL | fn test<T: Foo<i32, Assoc = i32>>() {}
+   |            ^^^------------------ help: remove the unnecessary generics
+   |            |
+   |            expected 0 generic arguments
+   |
+note: trait defined here, with 0 generic parameters
+  --> $DIR/dont-probe-missing-item-name-2.rs:1:7
+   |
+LL | trait Foo {
+   |       ^^^
+
+error[E0220]: associated type `Assoc` not found for `Foo`
+  --> $DIR/dont-probe-missing-item-name-2.rs:8:21
+   |
+LL | fn test<T: Foo<i32, Assoc = i32>>() {}
+   |                     ^^^^^ associated type `Assoc` not found
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0107, E0220.
+For more information about an error, try `rustc --explain E0107`.

--- a/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-3.rs
+++ b/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-3.rs
@@ -1,0 +1,11 @@
+trait Trait {
+    fn method() -> impl Sized;
+}
+
+// Ensure that we don't try to probe the name of the RPITIT when looking for
+// fixes to suggest for the missing associated type's trait path below.
+
+fn foo() -> i32::Assoc {}
+//~^ ERROR ambiguous associated type
+
+fn main() {}

--- a/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-3.stderr
+++ b/tests/ui/impl-trait/in-trait/dont-probe-missing-item-name-3.stderr
@@ -1,0 +1,15 @@
+error[E0223]: ambiguous associated type
+  --> $DIR/dont-probe-missing-item-name-3.rs:8:13
+   |
+LL | fn foo() -> i32::Assoc {}
+   |             ^^^^^^^^^^
+   |
+help: if there were a trait named `Example` with associated type `Assoc` implemented for `i32`, you could use the fully-qualified path
+   |
+LL - fn foo() -> i32::Assoc {}
+LL + fn foo() -> <i32 as Example>::Assoc {}
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0223`.

